File: src/0347-Top-K-Frequent-Elements/0347.py
# class Solution:
#     def topKFrequent(self, nums, k):
#         """
#         :type nums: List[int]
#         :type k: int
#         :rtype: List[int]
#         """
#         from collections import Counter
#         return [item[0] for item in Counter(nums).most_common(k)] 

class Solution:
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """   
        import heapq
        count_list = dict()
        for i in nums:
            count_list[i] = count_list.get(i, 0) + 1
        
        p = list()
        result = list()
        for i in count_list.items():
            heapq.heappush(p, (-i[1], -i[0]))
            if len(p) > len(count_list) - k:
                _, val = heapq.heappop(p)
                result.append(-val)

        return result

if __name__ == '__main__':
    nums = [4,1,1,1,2,2,3]
    k = 2
    print(Solution().topKFrequent(nums, k))

File: src/0347-Top-K-Frequent-Elements/0347.go
package topkfrequentelements

func topKFrequent(nums []int, k int) []int {
	mp := make(map[int]int)
	for _, val := range nums {
		mp[val]++
	}
	str := make([][]int, len(nums)+1)
	for k, v := range mp {
		str[v] = append(str[v], k)
	}
	ans := make([]int, 0)
	for i := len(nums); i >= 1; i-- {
		if len(ans) >= k {
			break
		}
		ans = append(ans, str[i]...)
	}
	return ans
}
